Introduce NodeAutoComplete Feature

QilongTang opened the pull request:
Please Note:
1. Before submitting the PR, please review [How to Contribute to Dynamo](https://github.com/DynamoDS/Dynamo/blob/master/CONTRIBUTING.md)
2. Dynamo Team will meet 1x a month to review PRs found on Github (Issues will be handled separately)
3. PRs will be reviewed from oldest to newest
4. If a reviewed PR requires changes by the owner, the owner of the PR has 30 days to respond. If the PR has seen no activity by the next session, it will be either closed by the team or depending on its utility will be taken over by someone on the team
5. PRs should use either Dynamo's default PR template or [one of these other template options](https://github.com/DynamoDS/Dynamo/wiki/Choosing-a-Pull-Request-Template) in order to be considered for review.
6. PRs that do not have one of the Dynamo PR templates completely filled out with all declarations satisfied will not be reviewed by the Dynamo team.
7. PRs made to the `DynamoRevit` repo will need to be cherry-picked into all the DynamoRevit Release branches that Dynamo supports. Contributors will be responsible for cherry-picking their reviewed commits to the other branches after a `LGTM` label is added to the PR.

### Purpose

Draft PR, more to come.

![NodeAutocompleteCandidates_v3](https://user-images.githubusercontent.com/3942418/93265106-1c3a3600-f776-11ea-8bb2-72d0fdc5ddcb.gif)


### Declarations

Check these if you believe they are true

- [x] The codebase is in a better state after this PR
- [x] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [x] The level of testing this PR includes is appropriate
- [x] User facing strings, if any, are extracted into `*.resx` files
- [x] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [x] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions) and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.
- [ ] This PR modifies some build requirements and the readme is updated

### Reviewers

@DynamoDS/dynamo 

### FYIs

(FILL ME IN, Optional) Names of anyone else you wish to be notified of

QilongTang commented on src/DynamoCoreWpf/Controls/NodeAutoCompleteSearchControl.xaml.cs:
```diff
@@ -0,0 +1,256 @@
+    /// Notice this control shares a lot of logic with InCanvasSearchControl for now
+    /// But they will diverge eventually because of UI improvements to auto complete.
+    /// </summary>
+    public partial class NodeAutoCompleteSearchControl
```
WPF does not support that

aparajit-pratap replied:
Why not just reuse the `InCanvasSearchControl ` control in that case? If not, what are the differences between the two controls?

QilongTang replied:
The visibility of context menu, visibility of the control itself. The entry onclick behavior are all different and I expect more difference in the future

QilongTang replied:
The UI code is actually fairly different

aparajit-pratap replied:
I'm not sure I understand completely but will take your word for it. The entry onclick behavior, I agree, is different, but it is just a way to bring up this control, not the control itself.

QilongTang commented on src/DynamoCoreWpf/Controls/NodeAutoCompleteSearchControl.xaml.cs:
```diff
@@ -0,0 +1,256 @@
+
+        }
+
+        private void InCanvasSearchControl_Unloaded(object sender, RoutedEventArgs e)
```
Will do

QilongTang commented on src/DynamoCoreWpf/Controls/NodeAutoCompleteSearchControl.xaml.cs:
```diff
@@ -0,0 +1,256 @@
+
+        private void currentApplicationDeactivated(object sender, EventArgs e)
+        {
+            OnRequestShowInCanvasSearch(ShowHideFlags.Hide);
```
Will do

aparajit-pratap commented on src/DynamoCoreWpf/Commands/WorkspaceCommands.cs:
```diff
@@ -27,6 +27,7 @@ public partial class WorkspaceViewModel
         private DelegateCommand _unpauseVisualizationManagerUpdateCommand;
         private DelegateCommand _showHideAllGeometryPreviewCommand;
         private DelegateCommand _showInCanvasSearchCommand;
+        private DelegateCommand _showNodeAutoCompleteSearchCommand;
```
Can we rename these private fields to not have the C++ style of naming them?

QilongTang replied:
Updated in latest commit

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs:
```diff
@@ -364,6 +364,23 @@ private bool CanConnect(object parameter)
+
+        private bool CanAutoComplete(object parameter)
+        {
+            return true;
```
I think the check for `dynamoViewModel.EnableNodeAutoComplete` should be here as that condition dictates if `CanAutoComplete` is true or false.

QilongTang replied:
But we dont have access to DynamoViewModel here?

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Core/WorkspaceViewModel.cs:
```diff
@@ -221,14 +227,20 @@ public DynamoPreferencesData DynamoPreferences
+        /// ViewModel that is used in NodeAutoComplete feature in context menu and called by Shift+DoubleClick.
+        /// </summary>
+        [JsonIgnore]
+        public SearchViewModel NodePortAutoCompleteSearchViewModel { get; private set; }
```
How about simplifying this name to `AutoCompleteSearchViewModel`? The long name is quite confusing IMO.

QilongTang replied:
How about `NodeAutoCompleteSearchViewModel`, AutoComplete itself is not specific enough IMO

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs:
```diff
@@ -370,13 +370,15 @@ private bool CanConnect(object parameter)
-            dynamoViewModel.CurrentSpaceViewModel.ShowNodeAutoCompleteSearchCommand.Execute(ShowHideFlags.Show);
+            // Since NodeAutoCompleteCommand is not triggered from xml event trigger directly,
+            // CanAutoComplete will not be triggered automatically.
+            if (CanAutoComplete(parameter))
```
I don't see this pattern used for other commands where the `Can` method is being called explicitly so why make an exception in this case?

QilongTang replied:
@aparajit-pratap Please refer to the comment, all the other view commands here are directly triggered by WPF xml event trigger but this one is not. This command is triggered by code by calling execute(). Calling execute() does not call `CanAutoComplete` at all. That is why my original change worked but after moving the logic to `CanAutoComplete`, it regressed. We can consider removing this command and simply implement a function as an alternative.

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Core/StateMachine.cs:
```diff
@@ -802,6 +812,16 @@ internal bool HandlePortClicked(PortViewModel portViewModel)
                 {
+                    if (Keyboard.Modifiers == ModifierKeys.Alt)
+                    {
+                        portViewModel.NodeAutoCompleteCommand.Execute(portViewModel);
```
@QilongTang I think we should use `if(portViewModel.NodeAutoCompleteCommand.CanExecute(portViewModel))` as a check here. This will make a call to `CanAutoComplete`. That's how the `DelegateCommand` pattern is supposed to work.

QilongTang replied:
Make sense! Will update

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs:
```diff
@@ -364,14 +367,37 @@ private bool CanConnect(object parameter)
+                DynamoViewModel dynamoViewModel = _node.DynamoViewModel;
+                var svm = dynamoViewModel.CurrentSpaceViewModel.NodeAutoCompleteSearchViewModel as NodeAutoCompleteSearchViewModel;
+                svm.PortViewModel = parameter as PortViewModel;
+                dynamoViewModel.CurrentSpaceViewModel.ShowNodeAutoCompleteSearchCommand.Execute(ShowHideFlags.Show);
```
Why can't we make `OnRequestNodeAutoCompleteSearch` internal and call it directly from here instead of defining yet another command (`ShowNodeAutoCompleteSearchCommand`) here? Something like:
```
dynamoViewModel.CurrentSpaceViewModel.OnRequestNodeAutoCompleteSearch(ShowHideFlags.Show);
```

aparajit-pratap replied:
I find too many "commands" is confusing.

QilongTang replied:
Will look into this

QilongTang replied:
@aparajit-pratap it works and I could remove the command `ShowNodeAutoCompleteSearchCommand` on workspaceViewModel layer. We may want to add recordable command support in the future and look at analytics.. But maybe not needed as a starting point?

aparajit-pratap replied:
The reason why I thought this command was redundant is since we already have `NodeAutoCompleteCommand`, which I think in turn was calling this command, correct me if I'm wrong.

QilongTang replied:
@aparajit-pratap They are two different commands on different view model. But I have removed the one on workspace view model already.

aparajit-pratap commented on test/DynamoCoreWpfTests/NodeAutoCompleteSearchTests.cs:
```diff
@@ -0,0 +1,121 @@
+            type = port.GetInputPortType();
+            Assert.AreEqual("string", type);
+
+            var searchViewModel = (ViewModel.CurrentSpaceViewModel.NodeAutoCompleteSearchViewModel as NodeAutoCompleteSearchViewModel);
```
I was hoping we could instantiate a new `NodeAutoCompleteSearchViewModel` with a new `PortViewModel` each time we wanted autocomplete, something like:
```
var searchViewModel = new NodeAutoCompleteSearchViewModel(inPorts[1]);
var suggestions = searchViewModel.GetMatchingNodes();
```
as I thought from a usability standpoint of the API this looks cleaner but unfortunately, we can't do that as it is currently designed I suppose.

QilongTang replied:
hmm since the base class of `NodeAutoCompleteSearchViewModel` take DynamoViewModel from constructor, it is currently not doable but we could write an alternative constructor takes both. Well, I guess I am curious if you see other benefits of initializing it more than doing it only once?

aparajit-pratap replied:
The only benefit I see is from a code clarity standpoint. It is a trade-off between:
1. Use a single instance of `NodeAutoCompleteSearchViewModel` and replace its contents each time by accessing the same object each time from the workspace model vs
2. Create a new temporary object each time, use it for autocomplete where needed and forget about it

QilongTang replied:
I agree. Since we adopted a majority code of InCanvasSearchViewModel which is a singleton, it may make sense to touch how this works later? If it is not urgent.. What do you think?

aparajit-pratap commented on src/DynamoCoreWpf/ViewModels/Core/PortViewModel.cs:
```diff
@@ -364,14 +364,34 @@ private bool CanConnect(object parameter)
+        private void AutoComplete(object parameter)
+        {
+            // Since NodeAutoCompleteCommand is not triggered from xml event trigger directly,
+            // CanAutoComplete will not be triggered automatically.
```
I think we can cleanup these comments :)

QilongTang commented:
@aparajit-pratap @mmisol @reddyashish  @reddyashish Any other comments I can address for you?

aparajit-pratap commented on src/DynamoCoreWpf/Controls/NodeAutoCompleteSearchControl.xaml.cs:
```diff
@@ -0,0 +1,258 @@
+    {
+        ListBoxItem HighlightedItem;
+
+        internal event Action<ShowHideFlags> RequestShowNodeAutoCompleteSearch;
```
@QilongTang where is this event registered? How does it work?

QilongTang replied:
https://github.com/DynamoDS/Dynamo/blob/NodeAutoComplete/src/DynamoCoreWpf/Controls/NodeAutoCompleteSearchControl.xaml.cs#L58.

aparajit-pratap replied:
That is the call but what method is it registered with? Which method gets called eventually?

aparajit-pratap replied:
I don't think it does anything currently.

QilongTang replied:
@aparajit-pratap Sorry it is registered in a different file, see: https://github.com/DynamoDS/Dynamo/blob/NodeAutoComplete/src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml.cs#L110

QilongTang replied:
This need to be done on workspace level because we are monitoring all the left click on WS level, so when user left click on anywhere else will dismiss the node autocomplete search bar

aparajit-pratap replied:
How is this different from `RequestNodeAutoCompleteSearch` defined on `WorkspaceViewModel`?

QilongTang replied:
Sry been so long since I wrote this, I debugged and believe it is leveraged as event driven here : https://github.com/DynamoDS/Dynamo/blob/NodeAutoComplete/src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml#L355

QilongTang replied:
My memory faded so my answer may not be clear before.. There are several kinds of show/hide request handled by different events on different level
1. request to show by alt+ left click
2. request to hide by left click on canvas
3. request to hide when window deactivated, e.g. user switched back to Revit, etc

These requests and events was handled on different level for in-canvas search so I inherited the same architecture, hopefully this answers your questions

aparajit-pratap commented on src/DynamoCoreWpf/Controls/NodeAutoCompleteSearchControl.xaml.cs:
```diff
@@ -0,0 +1,258 @@
+        {
+            var membersListBox = sender as ListBox;
+            // As soon as listbox renders, select first member.
+            membersListBox.ItemContainerGenerator.StatusChanged += OnMembersListBoxIcgStatusChanged;
```
What does the name mean - Icg status?

QilongTang replied:
This is from the legacy code for the control we are mimicing, it means `ItemContainerGenerator`

QilongTang commented:
@aparajit-pratap Let me know if you have more thoughts

aparajit-pratap commented on src/DynamoCoreWpf/Views/Core/WorkspaceView.xaml:
```diff
@@ -345,6 +345,16 @@
+               Placement="MousePoint"
+               HorizontalOffset="-290"
+               DataContext="{Binding NodeAutoCompleteSearchViewModel}">
+            <ui:NodeAutoCompleteSearchControl RequestShowNodeAutoCompleteSearch="ShowHideInCanvasControl" />
```
Based on your comments above, shouldn't this be `ShowHideNodeAutoCompleteControl`?

QilongTang replied:
@aparajit-pratap Yup, let me update

aparajit-pratap reviewed (commented):
@QilongTang just one last question and then I think this looks good for now. Are you going to merge this into master?

QilongTang commented:
@aparajit-pratap Addressed your comments! No wonder the window deactivate -> hide node autocomplete search bar workflow did not work for me.. Yes I intend to merge to master, since this is guarded by the preference setting and default to false without introducing new dll to our core delivery, I think these changes should be safe.

QilongTang commented:
All tests passed, merging
